src/Psalm/Internal/Codebase/Populator.php

```diff
@@ -509,6 +509,14 @@ private function populateDataFromParentClass(
             return;
         }
 
+        if ($parent_storage->inheritors) {
```

Comment:
I'm not sure what invalid_inheritors and active_inheritors means here.

Can't you just have each storage know what allowed_inheritors it declared and check, for each class if it respects inheritors declared by every parent_class/parent_interface?

Reply:
This was my naïve approach to dealing with with parents of parents, but I realise now that the `class_storage->parent_classes` deals with this for me. Much simpler now.